### ISSUE TITLE
"Fall through Error" "ngx_rtmp_eval.c" 160:17; Ubuntu 18.10 Nginx 10.15.5

### ISSUE THREAD

**Issue by Ericarthurc** [sergey-dryabzhinsky/nginx-rtmp-module#285]
I can't make nginx with the module do to an error in ngx_rtmp_eval.c; I am grabbing from the latest dev branch. And I can't seem to bypass the warnings.? I tried "make -Werror" but that didn't work.

**Comment by winshining** [comment]
Add

    /* fall through */

befre

    case ESCAPE:

will eliminate the error.

**Comment by jstilwell** [comment]
I'm having the same problem - above solution  not working for me.

**Comment by thepurplepainter** [comment]
Yes, I was having the same issue, as we have nested switches, and more than one fall through comment is needed. 

This article: [https://dzone.com/articles/implicit-fallthrough-in-gcc-7](https://dzone.com/articles/implicit-fallthrough-in-gcc-7) indicates fall-through comments must be placed in switch statements where the case conditions only cover expected values (defined explicitly, to ensure the compiler knows to move on (or "fall through") to the next line.

> The range and shape of “falls through” comments accepted are contingent upon the level of the warning. (The default level is =3.)
> 
> -Wimplicit-fallthrough=0 disables the warning altogether.
> -Wimplicit-fallthrough=1 treats any kind of comment as a “falls through” comment.
> -Wimplicit-fallthrough=2 essentially accepts any comment that contains something that matches (case insensitively) “falls?[ \t-]*thr(ough|u)” regular expression.
> -Wimplicit-fallthrough=3 case sensitively matches a wide range of regular expressions, listed in the GCC manual. E.g., all of these are accepted:/* Falls through. *//* fall-thru *//* Else falls through. *//* FALLTHRU *//* … falls through … */etc.
> -Wimplicit-fallthrough=4 also, case sensitively matches a range of regular expressions but is much more strict than level =3.
> -Wimplicit-fallthrough=5 doesn’t recognize any comments.
> Note that the attributes are recognized on any level.
> 
> For a more detailed description of the regular expressions accepted, please see the GCC manual.
> These “falls through” comments are meant to be used as in the example here:
> 
```
switch (cond)
 {
 case 0:
   foo (0);
   /* FALLTHRU */
 case 1:
   foo (1); 
   /* further code */
 }
```
> 
> They should precede the case or default keywords.

The article states that, given a switch with only 2 explicit cases, a fall through comment must be added before the 2nd case. If other values could exist, a break statement must be added after the 2nd explicit case to ensure the compiler continues to move on through the code. 

The issue we are seeing  occurs within nested switches, so I examined both the switch where the error is reported (`switch (c) {` is line 160) and its parent switch. I added the two fall-through comments and `default: break;` (sorry I wasn't able to figure out how to highlight this using Markdown)

```
            case NORMAL:
                switch (c) {
                    case '$':
                        name.data = p + 1;
                        state = NAME;
                        continue;
                    case '\\':
                        state = ESCAPE;
                        continue;
                    /* fall through */
                    default:
                        break;
                }
           /* fall through */
           case ESCAPE:
```

This change to **ngx_rtmp_eval.c** corrected the issue and I was able to make and install the package. I used instructions on the following site as a guide for building and installing nginx with RTMP module on my pi (running ArchArm64): 

[https://obsproject.com/forum/resources/how-to-set-up-your-own-private-rtmp-server-using-nginx.50/](https://obsproject.com/forum/resources/how-to-set-up-your-own-private-rtmp-server-using-nginx.50/)

**Comment by thepurplepainter** [comment]
Same here. Maybe I just didn't understand the instructions above... I dug around on the interwebz and found this article: [https://dzone.com/articles/implicit-fallthrough-in-gcc-7](https://dzone.com/articles/implicit-fallthrough-in-gcc-7) which indicates fall-through comments must be placed in switch statements to ensure the compiler will allow other values than those defined explicitly.

> These “falls through” comments are meant to be used as in the example here:
> 
```
> switch (cond)
>  {
>  case 0:
>    foo (0);
>    /* FALLTHRU */
>  case 1:
>    foo (1); 
>    /* further code */
>  }
```
> They should precede the case or default keywords. Be aware, though, that they might clash with macros so sometimes using the attributes will be necessary.

Based on this, I updated as follows (where `switch (c) {` is line 160 and `case ESCAPE:` was originally line 170 and my additions are in bold):

```
            case NORMAL:
                switch (c) {
                    case '$':
                        name.data = p + 1;
                        state = NAME;
                        continue;
                    case '\\':
                        state = ESCAPE;
                        continue;
                    /* fall through */
                    default:
                        break;
                }
           /* fall through */
           case ESCAPE:
```

This change to **ngx_rtmp_eval.c** corrected the issue and I was able to make and install the package. I used instructions on the following site as a guide for building and installing nginx with RTMP module on my pi (running ArchArm64): 

https://obsproject.com/forum/resources/how-to-set-up-your-own-private-rtmp-server-using-nginx.50/

I'm still trying to get the RTMP part to work (maybe not compatible with the most recent nginx, not sure), but the compile errors are gone...

**Comment by sundbry** [comment]
You can work around this without patching by configuring NGINX with ```--with-cc-opt="-Werror=implicit-fallthrough=0 ``` to ignore the error in GCC

**Comment by sergey-dryabzhinsky** [comment]
All fixes from `arut` and suggested in `dev`-branch. You can try it again.
